components/pixels/lib.rs

```diff
@@ -395,10 +410,90 @@ pub struct ImageMetadata {
     pub height: u32,
 }
 
+/// The images store their bytes in this storage.
+#[derive(Debug, Default)]
+pub struct ImageByteStorage {
+    storage: RwLock<Vec<Vec<u8>>>,
+}
+
+/// Image cache depends on clone being a deep clone.
+impl Clone for ImageByteStorage {
+    fn clone(&self) -> Self {
+        let vector = self.storage.read().unwrap().to_owned();
+        Self {
+            storage: RwLock::new(vector),
+        }
+    }
+}
+
+impl MallocSizeOf for ImageByteStorage {
+    fn size_of(&self, ops: &mut malloc_size_of::MallocSizeOfOps) -> usize {
+        self.storage
+            .read()
+            .unwrap()
+            .iter()
+            .map(|s| s.size_of(ops))
+            .sum()
+    }
+}
+
+impl<'a> ImageByteStorage {
+    pub fn reader(&'a self) -> ImageByteStorageReader<'a> {
+        ImageByteStorageReader(self.storage.read().unwrap())
+    }
+
+    pub fn writer(&'a self) -> ImageByteStorageWriter<'a> {
+        ImageByteStorageWriter(self.storage.write().unwrap())
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, MallocSizeOf)]
+pub struct ImageByteStorageIndex(usize);
+
+impl ImageByteStorageIndex {
+    #[allow(unused)]
+    pub fn mock_for_testing() -> Self {
+        ImageByteStorageIndex(0)
+    }
+}
+
+pub trait ImageByteStorageRead {
+    fn get(&self, index: &ImageByteStorageIndex) -> Option<&Vec<u8>>;
+}
+
+#[derive(Debug)]
+pub struct ImageByteStorageReader<'a>(RwLockReadGuard<'a, Vec<Vec<u8>>>);
+
+impl<'a> ImageByteStorageRead for ImageByteStorageReader<'a> {
+    fn get(&self, index: &ImageByteStorageIndex) -> Option<&Vec<u8>> {
+        self.0.get(index.0)
+    }
+}
+
+#[derive(Debug)]
+pub struct ImageByteStorageWriter<'a>(RwLockWriteGuard<'a, Vec<Vec<u8>>>);
+
+impl<'a> ImageByteStorageWriter<'a> {
+    pub fn insert(&mut self, data: Vec<u8>) -> ImageByteStorageIndex {
+        let index = self.0.len();
+        self.0.push(data);
+        ImageByteStorageIndex(index)
+    }
+}
+
+impl<'a> ImageByteStorageRead for ImageByteStorageWriter<'a> {
+    fn get(&self, index: &ImageByteStorageIndex) -> Option<&Vec<u8>> {
+        self.0.get(index.0)
+    }
+}
+
 // FIXME: Images must not be copied every frame. Instead we should atomically
 // reference count them.
-
-pub fn load_from_memory(buffer: &[u8], cors_status: CorsStatus) -> Option<RasterImage> {
+pub fn load_from_memory(
+    buffer: &[u8],
+    cors_status: CorsStatus,
+    storage: &mut ImageByteStorageWriter,
+) -> Option<RasterImage> {
```

Comment:
Why pass in the storage? Why not just assign an `Arc<Vec<u8>>` to the new `RasterImage`?